### PR TITLE
[mlir][vector] Add deinterleave operation to vector dialect

### PR DESCRIPTION
**Pull request by mub-at-arm**
The deinterleave operation constructs two vectors from a single input
vector. The first result vector contains the elements from even indexes
of the input, and the second contains elements from odd indexes. This is
the inverse of a `vector.interleave` operation.

Each output's trailing dimension is half of the size of the input
vector's trailing dimension. This operation requires the input vector
to have a rank > 0 and an even number of elements in its trailing
dimension.

The operation supports scalable vectors.

Example:
```mlir
%0, %1 = vector.deinterleave %a
           : vector<8xi8> -> vector<4xi8>
%2, %3 = vector.deinterleave %b
           : vector<2x8xi8> -> vector<2x4xi8>
%4, %5 = vector.deinterleave %c
           : vector<2x8x4xi8> -> vector<2x8x2xi8>
%6, %7 = vector.deinterleave %d
           : vector<[8]xf32> -> vector<[4]xf32>
%8, %9 = vector.deinterleave %e
           : vector<2x[6]xf64> -> vector<2x[3]xf64>
%10, %11 = vector.deinterleave %f
           : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>
```

### CODE REVIEW

File: mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir [hunk 1]
```diff
@@ -942,6 +942,17 @@ llvm.func @vector_insert_extract(%f256: vector<8xi32>, %f128: vector<4xi32>,
...
 
+// CHECK-LABEL: @vector_deinterleave2
+llvm.func @vector_deinterleave2(%vec1: vector<4xf64>, %vec2: vector<[8]xi32>) {
+  // CHECK: call { <2 x double>, <2 x double> } @llvm.vector.deinterleave2.v4f64(<4 x double> %{{.*}})
+  %0 = "llvm.intr.vector.deinterleave2" (%vec1) :
+              (vector<4xf64>) -> !llvm.struct<(vector<2xf64>, vector<2xf64>)>
+  // CHECK: call { <vscale x 4 x i32>, <vscale x 4 x i32> } @llvm.vector.deinterleave2.nxv8i32(<vscale x 8 x i32> %{{.*}})
+  %1 = "llvm.intr.vector.deinterleave2" (%vec2) :
+              (vector<[8]xi32>) -> !llvm.struct<(vector<[4]xi32>, vector<[4]xi32>)>
+  llvm.return
```
Comment (MacDue) [inline mlir/test/Target/LLVMIR/llvmir-intrinsics.mlir:953]: It looks like you need to rebase this PR, looks like it contains some previous changes.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 2]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+    }]>>,
+    ResultIsHalfSourceVectorType<"res1">,
+    ResultIsHalfSourceVectorType<"res2">,
+    AllTypesMatch<["res1", "res2"]>
+    ]> {
+      let summary = "constructs two vectors by deinterleaving an input vector";
+      let description = [{
+        The deinterleave operation constructs two vectors from a single input
+        vector. Each new vector is the collection of even and odd elements
+        from the input, respectively. This is essentially the inverse of an
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:579]: nit: I'd drop "essentially" (this is the inverse).
```suggestion
        from the input, respectively. This is the inverse of an
```

Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:579]: I think this could be clear:
```suggestion
        vector. The first result vector contains the elements from even 
        lanes of the input, and the second contains elements from odd lanes.
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 3]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        interleave operation.
+
+        Each output's size is half of the input vector's trailing dimension
+        for the n-D case and only dimension for 1-D cases. It is not possible
+        to conduct the operation on 0-D inputs or vectors where the size of
+        the (trailing) dimension is 1.
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:585]: Maybe reword a little?:
```suggestion
        The trailing dimension of each output is half of the size of input vector's 
        trailing dimension. This operation requires that the input vector has a 
        rank > 0 and a even number of elements its trailing dimension.
```

**Review by MacDue — COMMENTED** [review]
A few comments on the documentation:

**Comment by MacDue** [comment]
Marked as draft due to the rebase issue. I'll help @mub-at-arm resolve the issues with this PR tomorrow.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 4]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        %1 = vector.deinterleave %b
+                   : vector<8xi8>        ; yields vector<4xi8>, vector<4xi8>
+        %2 = vector.deinterleave %c
+                   : vector<2x8xf32>     ; yields vector<2x4xf32>, vector<2x4xf32>
+        %3 = vector.deinterleave %d
+                   : vector<2x4x[6]xf64> ; yields vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
+        ```
+      }];
+
+      let arguments = (ins AnyVector:$input);
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:602]: nit: Rename `$input` to `$source` (and update the getters/references accordingly), that'd be more consistent with other operations in the vector dialect. 
```suggestion
      let arguments = (ins AnyVector:$source);
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 5]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
   }];
 }
 
+class ResultIsHalfSourceVectorType<string result> : TypesMatchWith<
+  "type of 'input' is double the width of results",
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:547]: nit: specify this checks the trailing dimension?
```suggestion
  "the trailing dimension of the results is half the width of source trailing dimension",
```

File: mlir/test/Dialect/Vector/invalid.mlir [hunk 6]
```diff
@@ -1798,3 +1798,59 @@ func.func @invalid_outerproduct1(%src : memref<?xf32>) {
...
+  return
+}
+
+// -----
+
+func.func @deinterleave_scalable_rank_fail(%vec : vector<2x[4]xf32>) {
+  // expected-error @+1 {{'vector.deinterleave' op failed to verify that type of 'input' is double the width of results}}
+  %0, %1 = "vector.deinterleave" (%vec) : (vector<2x[4]xf32>) -> (vector<2x[2]xf32>, vector<[2]xf32>)
+  return
+}
```
Comment (MacDue) [inline mlir/test/Dialect/Vector/invalid.mlir:1856]: very minor nit: Can you ensure your editor adds newlines at the end of files (and update two test files)?
See [here](https://stackoverflow.com/questions/44704968/visual-studio-code-insert-newline-at-the-end-of-files) for a guide on this for vscode.

**Review by MacDue — COMMENTED** [review]
a few more nits:

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 7]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        Example:
+        ```mlir
+        %0 = vector.deinterleave %a
+                   : vector<[4]xi32>     ; yields vector<[2]xi32>, vector<[2]xi32>
+        %1 = vector.deinterleave %b
+                   : vector<8xi8>        ; yields vector<4xi8>, vector<4xi8>
+        %2 = vector.deinterleave %c
+                   : vector<2x8xf32>     ; yields vector<2x4xf32>, vector<2x4xf32>
+        %3 = vector.deinterleave %d
+                   : vector<2x4x[6]xf64> ; yields vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:598]: The docs should show this returns two results (I don't think assigning it to one value is valid): 
```suggestion
        %0, %1 = vector.deinterleave %a
                   : vector<[4]xi32>     ; yields vector<[2]xi32>, vector<[2]xi32>
        %2, %3 = vector.deinterleave %b
                   : vector<8xi8>        ; yields vector<4xi8>, vector<4xi8>
        %4, %5 = vector.deinterleave %c
                   : vector<2x8xf32>     ; yields vector<2x4xf32>, vector<2x4xf32>
        %6, %7 = vector.deinterleave %d
                   : vector<2x4x[6]xf64> ; yields vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 8]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        vector's trailing dimension. This operation requires the input vector
+        to have a rank > 0 and an even number of elements in its trailing
+        dimension.
+
+        The operation supports scalable vectors.
+
+        Example:
+        ```mlir
+        %0 = vector.deinterleave %a
+                   : vector<[4]xi32>     ; yields vector<[2]xi32>, vector<[2]xi32>
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:592]: What do you think about adding the result type? I think this will make it more readable but not too verbose, and we won't have to add comments like this
```mlir
%0, %1 = vector.deinterleave %a : vector<[4]xi32> into vector<[2]xi32>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 9]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+      let extraClassDeclaration = [{
+        VectorType getSourceVectorType() {
+          return ::llvm::cast<VectorType>(getInput().getType());
+        }
+        VectorType getResultOneVectorType() {
+          return ::llvm::cast<VectorType>(getRes1().getType());
+        }
+        VectorType getResultTwoVectorType() {
+          return ::llvm::cast<VectorType>(getRes2().getType());
+        }
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:618]: Aren't these the same?

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 10]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+    ResultIsHalfSourceVectorType<"res1">,
+    ResultIsHalfSourceVectorType<"res2">,
+    AllTypesMatch<["res1", "res2"]>
+    ]> {
+      let summary = "constructs two vectors by deinterleaving an input vector";
+      let description = [{
+        The deinterleave operation constructs two vectors from a single input
+        vector. The first result vector contains the elements from even lanes
+        of the input, and the second contains elements from odd lanes. This is
+        the inverse of a 'vector.interleave' operation.
```
[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:580]: 'lane' in the `vector.mask` documentation refers to the same concept in interleave/deinterleave.

[the remark this answers is not in the document]
Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:580]: but you could say "elements from even indexes..." instead of lanes

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td (hunk 9 above)
Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:618]: They are -- there should be a single  `getResultVectorType()` method +1

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 11]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+                   : vector<2x4x[6]xf64> into vector<2x4x[3]xf64>, vector<2x4x[3]xf64>
+        ```
+      }];
+
+      let arguments = (ins AnyVector:$source);
+      let results = (outs AnyVector:$res1, AnyVector:$res2);
+
+      let assemblyFormat = [{
+        $source attr-dict `:` type($source)
+      }];
```
Comment (MacDue) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:610]: The examples and `assemblyFormat` don't match. Also, I think @kuhar's suggestion was just to show one of the result types (since the other is the same). 

e.g.:
```
$source attr-dict `:` type($source) `into` type($res1)
```

Comment (mub-at-arm) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: @MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.

So the resulting IR should look like this:
`%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: That looks good to me :+1:

**Review by MacDue — APPROVED** [review]
This LGTM, though I'll wait to see if @dcaballe or @nicolasvasilache have any comments :slightly_smiling_face:

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td (hunk 11 above)
Comment (c-rhodes) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: I'm not a fan of that assembly format, I think consistency is much more important and would rather have the verbose form.

Comment (MacDue) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: The verbose form is not consistent with `vector.interleave` (which only shows the source type) -- (similar to `vectpr.shuffle`)

Comment (c-rhodes) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: I meant consistency in general not in relation to interleave.

Comment (mub-at-arm) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: Should it be consistent to interleave or the other MLIR operation formats?

Comment (kuhar) [reply mlir/include/mlir/Dialect/Vector/IR/VectorOps.td]: > @MacDue, @kuhar, I've added and `assemblyFormat` that fully describes the operation without being too verbose. An "x2" suffix on the result describes the quantity of that type quite simply and should be easily readable too.
> 
> So the resulting IR should look like this: `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>x2`

I'd prefer just `%1, %2 = vector.deinterleave %a : vector<4xf32> -> vector<2xf32>` for readability and consistency with most of core mlir dialects.
Separately, I think we should also have the result type on `vector.interleave` (I wasn't pinged on the initial PR IIRC, and I didn't catch it before it landed).

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 12]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        vector's trailing dimension. This operation requires the input vector
+        to have a rank > 0 and an even number of elements in its trailing
+        dimension.
+
+        The operation supports scalable vectors.
+
+        Example:
+        ```mlir
+        %0, %1 = vector.deinterleave %a
+                   : vector<[4]xi32> -> vector<[2]xi32>x2
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:595]: nit: I think it would be more digestible for the readers to lead with easier examples, i.e., simple static 1d vector, and move n-d vectors and scalable vectors towards the bottom.

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 13]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+        %6, %7 = vector.deinterleave %d
+                   : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>x2
+        ```
+      }];
+
+      let arguments = (ins AnyVector:$source);
+      let results = (outs AnyVector:$res1, AnyVector:$res2);
+
+      let assemblyFormat = [{
+        $source attr-dict `:` type($source) `->` type($res1) `` `x2`
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:609]: I don't think we need the `x2`, see: https://github.com/llvm/llvm-project/pull/92409#discussion_r1608455010

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 14]
```diff
@@ -543,6 +543,82 @@ def Vector_InterleaveOp :
...
+    ResultIsHalfSourceVectorType<"res1">,
+    ResultIsHalfSourceVectorType<"res2">,
+    AllTypesMatch<["res1", "res2"]>
+    ]> {
+      let summary = "constructs two vectors by deinterleaving an input vector";
+      let description = [{
+        The deinterleave operation constructs two vectors from a single input
+        vector. The first result vector contains the elements from even indexes
+        of the input, and the second contains elements from odd indexes. This is
+        the inverse of a 'vector.interleave' operation.
```
Comment (kuhar) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:583]:
```suggestion
        the inverse of a `vector.interleave` operation.
```

**Review by kuhar — COMMENTED** [review]
A few cosmetic issues, otherwise looks good

**Review by kuhar — APPROVED** [review]

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 15]
```diff
@@ -543,6 +543,86 @@ def Vector_InterleaveOp :
...
+>;
+
+def SourceVectorEvenElementCount : PredOpTrait<
+  "the trailing dimension of the source vector has an even number of elements",
+    CPred<[{
+      [&](){
+        auto srcVec = getSourceVectorType();
+        return srcVec.getDimSize(srcVec.getRank() - 1) % 2 == 0;
+      }()
+    }]>
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:569]: nit: fix indentation

```suggestion
  CPred<[{
    [&](){
      auto srcVec = getSourceVectorType();
      return srcVec.getDimSize(srcVec.getRank() - 1) % 2 == 0;
    }()
  }]>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 16]
```diff
@@ -543,6 +543,86 @@ def Vector_InterleaveOp :
...
+        vector's trailing dimension. This operation requires the input vector
+        to have a rank > 0 and an even number of elements in its trailing
+        dimension.
+
+        The operation supports scalable vectors.
+
+        Example:
+        ```mlir
+        %0, %1 = vector.deinterleave %a
+                   :vector<8xi8> -> vector<4xi8>
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:595]: nit: space
```suggestion
                   : vector<8xi8> -> vector<4xi8>
```

File: mlir/include/mlir/Dialect/Vector/IR/VectorOps.td [hunk 17]
```diff
@@ -543,6 +543,86 @@ def Vector_InterleaveOp :
...
+        %2, %3 = vector.deinterleave %b
+                   : vector<2x8xi8> -> vector<2x4xi8>
+        %4, %5 = vector.deinterleave %b
+                   : vector<2x8x4xi8> -> vector<2x8x2xi8>
+        %6, %7 = vector.deinterleave %c
+                   : vector<[8]xf32> -> vector<[4]xf32>
+        %8, %9 = vector.deinterleave %d
+                   : vector<2x[6]xf64> -> vector<2x[3]xf64>
+        %10, %11 = vector.deinterleave %d
+                   : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>
```
Comment (c-rhodes) [inline mlir/include/mlir/Dialect/Vector/IR/VectorOps.td:605]: nit: update var names
```suggestion
        %4, %5 = vector.deinterleave %c
                   : vector<2x8x4xi8> -> vector<2x8x2xi8>
        %6, %7 = vector.deinterleave %d
                   : vector<[8]xf32> -> vector<[4]xf32>
        %8, %9 = vector.deinterleave %e
                   : vector<2x[6]xf64> -> vector<2x[3]xf64>
        %10, %11 = vector.deinterleave %f
                   : vector<2x4x[6]xf64> -> vector<2x4x[3]xf64>
```

**Review by c-rhodes — APPROVED** [review]
Few minor nits but otherwise LGTM, cheers!

**Comment by MacDue** [comment]
If there's no further comments I'll land this tomorrow :slightly_smiling_face:

**Review by banach-space — APPROVED** [review]
